### DISCUSSION TITLE
Attributes to output data

### DISCUSSION THREAD

**Discussion by mrichter91** [synthetichealth/synthea#1429] [General]
Hello,

I'm currently implementing a custom module with the Synthea Module builder. Is there any way for me to export the attributes that I set in the module to the generated patient data somehow?

Thanks in advance for any hint!

Greetings!

**Comment by jawalonoski** [comment]
```json
{
  "name": "Attribute Example",
  "remarks": [
    "Example of recording an attribute in the patient record."
  ],
  "states": {
    "Initial": {
      "type": "Initial",
      "direct_transition": "Wait a Year"
    },
    "My Attribute": {
      "type": "SetAttribute",
      "attribute": "my_attribute",
      "direct_transition": "Encounter",
      "distribution": {
        "kind": "UNIFORM",
        "round": true,
        "parameters": {
          "high": 20,
          "low": 10
        }
      }
    },
    "Wait a Year": {
      "type": "Delay",
      "distribution": {
        "kind": "EXACT",
        "parameters": {
          "value": 1
        }
      },
      "unit": "years",
      "direct_transition": "My Attribute"
    },
    "Encounter": {
      "type": "Encounter",
      "reason": "",
      "telemedicine_possibility": "none",
      "direct_transition": "Record My Attribute",
      "wellness": true
    },
    "Record My Attribute": {
      "type": "Observation",
      "category": "survey",
      "unit": "",
      "codes": [
        {
          "system": "LOINC",
          "code": "1234",
          "display": "LOINC Code"
        }
      ],
      "direct_transition": "End Encounter",
      "attribute": "my_attribute"
    },
    "End Encounter": {
      "type": "EncounterEnd",
      "direct_transition": "Wait a Year"
    }
  },
  "gmf_version": 2
}
```

**Comment by mrichter91** [comment]
> ```json
> {
>   "name": "Attribute Example",
>   "remarks": [
>     "Example of recording an attribute in the patient record."
>   ],
>   "states": {
>     "Initial": {
>       "type": "Initial",
>       "direct_transition": "Wait a Year"
>     },
>     "My Attribute": {
>       "type": "SetAttribute",
>       "attribute": "my_attribute",
>       "direct_transition": "Encounter",
>       "distribution": {
>         "kind": "UNIFORM",
>         "round": true,
>         "parameters": {
>           "high": 20,
>           "low": 10
>         }
>       }
>     },
>     "Wait a Year": {
>       "type": "Delay",
>       "distribution": {
>         "kind": "EXACT",
>         "parameters": {
>           "value": 1
>         }
>       },
>       "unit": "years",
>       "direct_transition": "My Attribute"
>     },
>     "Encounter": {
>       "type": "Encounter",
>       "reason": "",
>       "telemedicine_possibility": "none",
>       "direct_transition": "Record My Attribute",
>       "wellness": true
>     },
>     "Record My Attribute": {
>       "type": "Observation",
>       "category": "survey",
>       "unit": "",
>       "codes": [
>         {
>           "system": "LOINC",
>           "code": "1234",
>           "display": "LOINC Code"
>         }
>       ],
>       "direct_transition": "End Encounter",
>       "attribute": "my_attribute"
>     },
>     "End Encounter": {
>       "type": "EncounterEnd",
>       "direct_transition": "Wait a Year"
>     }
>   },
>   "gmf_version": 2
> }
> ```

Thanks for the quick reply. This seems like the resulting JSON from the module builder, right?

I actually meant if there's a way to see this attribute in the resulting files after calling run_synthea with the module.
Sorry for the misunderstanding.

**Reply by jawalonoski** [reply]
That module code shows how to record the attribute into the health record using an Observation. If you don't record it, you won't see it.

Once recorded using an Observation, it will show up as appropriate given the export format you are using.

Alternatively, you can use the [JSON Export](https://github.com/synthetichealth/synthea/wiki/JSON-Export) to dump all the attribute values. But beware that this is a debugging format, and not any type of standard health data.

**Comment by mrichter91** [comment]
> That module code shows how to record the attribute into the health record using an Observation. If you don't record it, you won't see it.
> 
> Once recorded using an Observation, it will show up as appropriate given the export format you are using.
> 
> Alternatively, you can use the [JSON Export](https://github.com/synthetichealth/synthea/wiki/JSON-Export) to dump all the attribute values. But beware that this is a debugging format, and not any type of standard health data.

Ah I understand now. Thank you so much for taking the time :)

**Comment by mrichter91** [comment]
> That module code shows how to record the attribute into the health record using an Observation. If you don't record it, you won't see it.
> 
> Once recorded using an Observation, it will show up as appropriate given the export format you are using.
> 
> Alternatively, you can use the [JSON Export](https://github.com/synthetichealth/synthea/wiki/JSON-Export) to dump all the attribute values. But beware that this is a debugging format, and not any type of standard health data.

I just implemented this solution and it seems to work almost perfectly.

The problem is that I'm also using a keep-module. For testing reasons I'm exporting the attribute I want to record as an observation right before terminal. There's no other path to terminal as of now.

When I'm generating 1000 patients with the keep-module, I get 1000 encounters (1 per patient, that's correct), but only 752 observations with the set LOINC Code.

Am I overseeing something here?

**Comment by mrichter91** [comment]
Update: 

So I looked a little bit for the issue. Right after my Initial State I have a distributed transition to 4 different states (each distribution weighs 0.25). In each of those states, the attribute to export will be either set to 1, 2, 3 or 4. 
If I leave it like that, I always get 752 observations after data generation.

However, if I set the attribute to 0 BEFORE the distributed transition, I have 1000 observations (as wanted, but 248 of those have value 0 on the respective attribute) . The thing is, I really need the distribution between those values and a default value is no option.

**Reply by jawalonoski** [reply]
I can't debug your module for you.

Are the other 248 patients dead or alive? If patients die before getting to a specific point in your module, will they have an observation?

Is there some situation keeping those patients from arriving at the terminal state in your module? Perhaps a guard, delay, or loop?

Are you filtering the years of output? The default is yes, and the last 10 years of output. See the configuration setting  `exporter.years_of_history`; the description for that setting under https://github.com/synthetichealth/synthea/wiki/Common-Configuration#other-export-settings
